### ISSUE TITLE
Documentation of automatic thunking of Ltac2 notation arguments could be more enlightening

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#18013]
#### Coq Version 8.18 (from master)

`.../Ltac2.Notations.v` has `Ltac2 Notation do := do0.`
`do 10 (msg "x")` works as expected, but `do0 10 (msg "x")` gives an error.
I expected they would be equivalent and that both would work.
If there's an intentional difference, that should be documented.

![image](https://github.com/coq/coq/assets/1253341/15d09cf5-7d68-4b65-a2fa-f468ede4fa9f)

![image](https://github.com/coq/coq/assets/1253341/84dc9ea3-c28f-453c-8a7b-f4f2e32457b6)

```
From Ltac2 Require Import Ltac2.
From Ltac2 Require Import Message.
Ltac2 m s := Message.print (Message.of_string s).
Goal True.

do0 10 (m "x").
```

**Comment by ppedrot** [comment]
This *is* documented: https://coq.inria.fr/refman/proof-engine/ltac2.html#abbreviations.

**Comment by jfehrle** [comment]
It's not understandable from that documentation.

Humor me: what syntax is required to call `do0` directly?

This doesn't work, either (without any abbreviation):

![image](https://github.com/coq/coq/assets/1253341/4f9dfb4a-05a9-45e6-bbc3-d18bf3b1256a)

**Comment by SkySkimmer** [comment]
>Furthermore, in order to make this kind of construction useful in practice in an effectful language such as Ltac2, any syntactic argument to an abbreviation is thunked on-the-fly during its expansion.

The doc then includes an example.

**Comment by JasonGross** [comment]
> This doesn't work, either (without any abbreviation):

Replace `n ()` with `n` in the definition of `mydo0` and it will work fine.  Invoke `do0` as `do0 (fun () => 10) (fun () => m "x")` and it should work fine.

**Comment by jfehrle** [comment]
The documentation could be more enlightening, e.g.:
- a more realistic example (foo doesn't do anything useful)
- unlike Ltac1, you can't use this to create a simple alias, e.g.

```
Ltac2 do3 := do.
do 10 (m "x").
Fail do3 10 (m "x").
```

@JasonGross > Replace `n ()` with `n` in the definition of `mydo0` and it will work fine.

It would have to be invoked as `do0 10 (fun () => m "x")`

**Comment by JasonGross** [comment]
> It would have to be invoked as `do0 10 (fun () => m "x")`

Yes, there is no way around this.  Unlike Ltac1, evaluation is eager by default rather than lazy, and, as in OCaml, there is no implicit cast between lazy and eager values as there is in, e.g., Haskell.

**Comment by JasonGross** [comment]
@jfehrle do you want to prepare a PR with a documentation improvement?

**Comment by jfehrle** [comment]
Yes, I'll do that.  Not today, though.
